### PR TITLE
Refactor hull

### PR DESCRIPTION
**Pull request by agoss94**
Refactoring of commons-geometry-hull module. The ConvexHull interface has been relocated to commons-geometry-core and ConvexHull2D is now part of commons-geometry-euclidean with a builder that combines the logic of the Akl-Toussaint heuristic and the Monotone-Chain to build a convex hull.

Line 291-295 in ConvexHull2D has to be added to the original code in order to allow for collinear points to be considered in all cases. 

All necessary test have been moved to the euclidean module as well.

### CODE REVIEW

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+
+        /** Appends the given point to a collection of possible hull points, if and only
+         * if the given point is outside of a constructed quadrilateral of extreme properties.
+         *
+         * @param point a given point.
+         * @return this instance.
+         */
+        public Builder append(Vector2D point) {
+
+            //Checks if the given point supersedes one of the corners.
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:175]: Whitespace after `//`. Same again on the next comment below.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+            return quadrilateral;
+        }
+
+        /** Checks if the given point supersedes one of the corners. If it does the old
+         * corner is removed and the point added to the collection of points.
+         *
+         * @param point a given point.
+         */
+        private void checkCorners(Vector2D point) {
+            if (minX == null || point.getX() < minX.getX()) {
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:255]: Here we only require one null check. If minX is null then the other corners are also null and all can be set to the point:
```Java
if (minX == null) {
    minX = maxX = minY = maxY = point;
    candidates.add(point);
    return;
}
```
IIUC this condition will also be met if `candidate.isEmpty()`. However using the null check does indicate that for the rest of the method the corners are not null.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+     * takes to construct the convex hull afterwards, depending on the point
+     * distribution.
+     *
+     * @see <a href=
+     *      "http://en.wikipedia.org/wiki/Convex_hull_algorithms#Akl-Toussaint_heuristic">
+     *      Akl-Toussaint heuristic (Wikipedia)</a>
+     */
+    public static final class Builder {
+
+        /** Corner of triangle with minimal x coordinate. */
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:129]: Q. `triangle` to `quadrilateral`?

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+
+        /** Checks if the given point is located within the convex quadrilateral.
+         * @param point the point to check
+         * @param quadrilateralPoints the convex quadrilateral, represented by 4 points
+         * @return {@code true} if the point is inside the quadrilateral, {@code false} otherwise
+         */
+        private boolean insideQuadrilateral(final Vector2D point,
+                                                   final List<? extends Vector2D> quadrilateralPoints) {
+
+            Vector2D v1 = quadrilateralPoints.get(0);
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:281]: Since this walks a circle around the quadrilateral we can start anywhere. IIUC this can be changed to:
```java
final int size = quadrilateralPoints.size();
Vector2D v1 = quadrilateralPoints.get(size - 1);
Vector2D v2 = quadrilateralPoints.get(0);

// ...

for (int i = 1; i < size; i++) {
    v1 = v2;
    v2 = quadrilateralPoints.get(i);
```

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+            final List<Vector2D> upperHull = new ArrayList<>();
+            for (int idx = pointsSortedByXAxis.size() - 1; idx >= 0; idx--) {
+                final Vector2D p = pointsSortedByXAxis.get(idx);
+                updateHull(p, upperHull);
+            }
+
+            // concatenate the lower and upper hulls
+            // the last point of each list is omitted as it is repeated at the beginning of the other list
+            final List<Vector2D> hullVertices = new ArrayList<>(lowerHull.size() + upperHull.size() - 2);
+            for (int idx = 0; idx < lowerHull.size() - 1; idx++) {
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:365]: Can these loops be changed to omit the first point of each half:
```java
for (int idx = 1; idx < lowerHull.size(); idx++) {
```

**Review by aherbert — COMMENTED** [review]
Thanks for the update. I've suggested a few optimisation changes (untested). Please review and see if they make sense.

One issue is a loss of the ConvexHull2D commit history. The file does have 5 commits which all have a Geometry Jira ticket. Keeping the history allows the changes to be followed back to the original version. The test classes have been moved. It is only this file that is deleted and added. You should be able to: checkout the deleted file from a commit before it was deleted; use `git mv` to move the new file to avoid a clash; and then again to move the old file to the new location; replace the contents with what you have; and delete the temp new file.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:1]: Old file has been retrieved and moved to the new location as suggested.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+            }
+        }
+
+        /** Checks if the given point is located within the convex quadrilateral.
+         * @param point the point to check
+         * @param quadrilateralPoints the convex quadrilateral, represented by 4 points
+         * @return {@code true} if the point is inside the quadrilateral, {@code false} otherwise
+         */
+        private boolean insideQuadrilateral(final Vector2D point,
+                                                   final List<? extends Vector2D> quadrilateralPoints) {
```
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Fixed

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+        /** A precision context for comparing points. */
+        private final Precision.DoubleEquivalence precision;
+
+        /** Indicates if collinear points on the hull shall be present in the output.
+         * If {@code false}, only the extreme points are added to the hull.
+         */
+        private final boolean includeCollinearPoints;
+
+        /**Return a {@link Builder} instance configured with the given precision
```
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Fixed

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 1 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Fixed

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 2 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: The observation is correct I changed the code. Although the method becomes a bit longer this way we can save 3 of the four null checks as suggested.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 3 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Yes, that should obviously mean quadrilateral.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 4 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Yes, the change could be made here. More importantly it showed that I missed an edge case when collinear points should be included. I fixed that as well.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 5 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:372]: I am going to check that tomorrow .

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+            Vector2D p2 = it.next();
+            Vector2D p3;
+
+            Vector2D v1;
+            Vector2D v2;
+
+            while (it.hasNext()) {
+                p3 = it.next();
+
+                v1 = p1.vectorTo(p2);
```
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:461]: I am going to check that tomorrow .

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 5 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Changing which point to omit was no problem.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Did you mean to precompute v1? I did that an then made the assigment at the end of the loop which worked without a problem.

**Comment by agoss94** [comment]
All review marks have been completed. Please review all subsequent changes made.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (aherbert) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: I did mean precompute v2, then inside the loop shift v2 to v1 at the start and compute v2 again.

However you have precomputed v1, and shifted v2 to v1 at the end of the loop which is effectively the same (i.e. do not duplicate computing Vectors). I prefer your code as all the shifting of variables is in one place.

Note that you did not drop `p1 = p2` inside the loop (where p1 is now redundant).

I'll try and review fully in the next few days.

[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java]: Ah I see. I did miss the assignment. I deleted the line and renamed the variables to better fit the now altered code. Thanks for pointing out.

**Comment by darkma773r** [comment]
Hello! I plan on reviewing this soon, although I might not have time until this weekend. Thanks for the hard work here!

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+
+            // Checks if the given point supersedes one of the corners.
+            checkCorners(point);
+
+            // Only proceed if the quadrilateral is complete.
+            if (candidates.size() < 4) {
+                return this;
+            }
+
+            final List<Vector2D> quadrilateral = buildQuadrilateral(minY, maxX, maxY, minX);
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:183]: It may be useful to cache the quadrilateral. Modification of the `checkCorners` method can be done to remove the cached quadrilateral if the point supersedes one of the corners.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+            Vector2D first = it.next();
+            Vector2D p1 = it.next();
+            Vector2D v1 = first.vectorTo(p1);
+
+            Vector2D p2;
+            Vector2D v2;
+
+            while (it.hasNext()) {
+                p2 = it.next();
+
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:460]: Remove this empty line.

**Review by aherbert — COMMENTED** [review]
This is looking good. I note that it is possible to cache the quadrilateral. For larger clouds of points this may be useful when the quadrilateral is unchanged by many successive points.

Let's wait for a second opinion from Matt.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 9 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:183]: Or maybe get rid of the list at all. The way I see it. The purpose of the list is to count how many of the points are truly distinct and to go around all sides in an ordered manner. Since we only have 4 corners at all and only proceed if there are  at least 3 sides it would be possible to get rid of the list and just to count and then replace the loop with a method that checks either all three or four sides.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+                null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String toString() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(getClass().getSimpleName())
+            .append("[vertices= ")
+            .append(getVertices())
```
[the remark this answers is not in the document]
Comment (darkma773r) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:88]: That's probably a good idea. If we do this here, we should update (in another ticket) other classes that currently produce arbitrarily large `toString` representations, like `LinePath`. They should be consistent in the number of vertices printed before truncation.

**Comment by darkma773r** [comment]
This looks great! Thank you, @agoss94! I think the only things to address are the two points made above (toString and caching the quadrilateral) and the code coverage. There are several if statement branches in `ConvexHull2D` that are not hit by the tests, which makes me think that either the tests are not sufficiently hitting all of the edge cases or those statements are not needed.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:88]: I would suggest to refrain from the change for now and to open a new ticket for this one. Gilles does have a different opinion on this topic (https://issues.apache.org/jira/projects/GEOMETRY/issues/GEOMETRY-110?filter=allopenissues) and even if we implement the change all the other classes still need an update. Better to change all the classes in one ticket.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 9 above)
Comment (agoss94) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:183]: It isn't really a new approach, but maybe more the fact, that the loop has only 3-4 iterations which could be straight forward to code. Since we only check the sides [minY, maxX], [maxX, maxY], [maxY, minX], [minX, minY]  for any given point, whereas we skip any side if the two points align, which can happen only once. I guess one could implement this logic entirely without a loop. 
I guess the simplest way is still to just cache the list.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (darkma773r) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:88]: That sounds good to me.

**Comment by agoss94** [comment]
@darkma773r I added a few test and removed some if branches which seemed to be really unnecessary. Most branches were not hit because those checked for identical points, which is superfluous as the PointsSet already filtered thoses out. 
Other if branches were not needed because of the comparator which orders the points before constructing the lower and upper hull. This allows us to make assumtions on the calculated distance to previously added points. 
You may double check the logic in deleting these checks.

**Comment by darkma773r** [comment]
This looks good to me! @aherbert, any objections to merging this in (after a rebase, it looks like)?

**Comment by aherbert** [comment]
I think the cache of the quadrilateral can be improved. It only has to be rebuilt if the 4 points for the extrema have changed. Here the builder is recreating it each time. IIUC if the corners change then you can rebuild the quadrilateral and add to the candidates. The rest of the work in `append` is not relevant and you can simply return. This can be done by returning a `boolean` from `checkCorners`, e.g.

```java
public Builder append(Vector2D point) {
    // Checks if the given point supersedes one of the corners.
    if (checkCorners(point)) {
        buildQuadrilateral(minY, maxX, maxY, minX);
        return this;
    }

    // if the quadrilateral is not well formed, e.g. only 2 points, do not attempt to reduce
    if (quadrilateral.size() < 3) {
        // Point cannot yet be dismissed.
        candidates.add(point);
        return this;
    }

    // check all points if they are within the quadrilateral
    // in which case they can not be part of the convex hull
    if (!insideQuadrilateral(point, quadrilateral)) {
        candidates.add(point);
    }

    return this;
}
```

**Comment by agoss94** [comment]
>
Proposition has been implemented. Thanks for the detailed review :) I learned a lot. 
The branch has also been rebased.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+        /**
+         * Checks if the given point supersedes one of the corners. If it does the old
+         * corner is removed and the point added to the collection of points.
+         *
+         * @param point a given point.
+         * @return {@code true} if any of the corners changed as a result of the check,
+         *         {@code false} otherwise.
+         */
+        boolean checkCorners(Vector2D point) {
+            boolean hasBeenModified = false;
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:261]: Note: You do not require this flag until after your `min == null` check

**Comment by aherbert** [comment]
@darkma773r I have no further suggestion for this. If you are OK with the new API then I suggest merging.
